core: fix inconsistency in mobj and memory attributes handling

etienne-lms opened the pull request:
Issue found as part of https://github.com/OP-TEE/optee_os/pull/1348

etienne-lms commented:
Note: tested on qemu, with and without core pager.

jenswi-linaro commented on core/arch/arm/mm/mobj.c:
```diff
@@ -49,7 +49,7 @@ struct mobj *mobj_sec_ddr;
 struct mobj_phys {
 	struct mobj mobj;
-	enum buf_is_attr battr;
+	enum buf_is_attr battr; /* Defined by CORE_MEM_* in core_memprot.h */
```
Is this comment needed? `CORE_MEM_*` are values of `enum buf_is_attr`.

etienne-lms replied:
that's right!

jenswi-linaro commented on core/arch/arm/mm/mobj.c:
```diff
@@ -159,7 +160,22 @@ struct mobj *mobj_phys_alloc(paddr_t pa, size_t size, uint32_t cattr,
+		area_type = MEM_AREA_NSEC_SHM;
+		break;
+	default:
+		DMSG("can't allocate only requested area");
```
Remove "only".

etienne-lms replied:
ok. i'll replace with "can't allocate with specified attribute".

By the way, i added the `CORE_MEM_TEE_RAM` even if it is not yet used. You're fine with that or should it be added only once we start using mobjs in the 'core' memory ?

jforissier commented on core/arch/arm/include/mm/core_memprot.h:
```diff
@@ -45,14 +45,13 @@
-	CORE_MEM_NSEC_SHM,
-	CORE_MEM_EXTRAM,
-	CORE_MEM_INTRAM,
-	CORE_MEM_CACHED,
```
I am in favor of trying to better detect misuses of `enum buf_is_attr` vs. `enum teecore_memtypes`, but I think this is a bit fragile. How about using different ranges and adding a few `asserts()`? Like this more or less:

```C
enum buf_is_attr {
        CORE_MEM_FIRST = 20, /* to detect misuse of teecore_memtypes */
        CORE_MEM_SEC,
        CORE_MEM_NON_SEC,
        /* ... */
};

enum teecore_memtypes {
        MEM_AREA_NOTYPE = 0,
        MEM_AREA_TEE_RAM,
        /* ... */
        MEM_AREA_TYPE_LAST
};

void foo(enum buf_is_attr attr)
{
        assert(attr > CORE_MEM_FIRST);
        /* ... */
}


void bar(enum teecore_memtypes type)
{
        assert(type < MEM_AREA_TYPE_LAST);
        /* ... */
}
```

etienne-lms replied:
Indeed this PR is NOT a fix and this does not aim at fixing bad implementations. Type verifications a compile time could help, but i am not sure `enum` types are good at this.

I though about your solution but won't this decrease optims of the switch/case implementations, without providing a real bullet proof solution ?

I feel a good solution would be to fully review the memory area and attributes definition and handling. But i have nothing to propose...

Jens, your opinion ?

jenswi-linaro replied:
GCC has a flag `-Wenum-compare` which is supposed to catch errors like these. It's supposed to be enabled via the `-Wall` flag so I'm a bit surprised we're having this error. Perhaps it doesn't cover assignments?

I agree with Etienne, this may decrease some optimizations. I don't have a good idea for a bulletproof fix, the only I can think of is getting rid of one of the enums but that's a bit complicated and out of scope here.

jforissier replied:
It looks like `-Wenum-compare` does nothing when the enum is passed as a function argument :(
OK we can leave this alone for now

jforissier commented:
@etienne-lms please use the exact enum names in the description of commit "core: fix inconsistency in mobj and memory attributes handling", that is: `CORE_MEM_TEE_RAM`, `CORE_MEM_TA_RAM`, `CORE_MEM_NSEC_SHM`. Also, `s/indetification/identification/`. With this done:
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

etienne-lms commented:
commit comment fixed, rebased, tag applied.

@jforissier, I added this comment in the second commit. Feel free to give your feedback (and not merge) if you don't agree:
```
A better approach would be to rely on the source code compiler to
assert the use of the right enumerated types. But tests showed that
even recent gcc versions do not (yet) fully verify such usage.
```

jforissier commented:
@etienne-lms I'm fine with the comment. Let's wait for @jenswi-linaro's feedback. Also please fix the checkpatch warning ;)

etienne-lms commented:
damn' checkpatch. i always forget it !
thanks.

jenswi-linaro commented:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

etienne-lms commented:
rebased, tags applied.
